### PR TITLE
Make NetPortableProfileTable thread-safe and test

### PR DESCRIPTION
**Pull request by joelverhagen**
Fixes:
- https://github.com/NuGet/Home/issues/778
- NuGet/Home#2464

Add on top of @hickford's good work with https://github.com/NuGet/NuGet2/pull/46.

/cc @emgarten @hickford

### CODE REVIEW

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 1]
```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
...
+                // This setter is only for unit tests and is NOT thread-safe.
+                // Reset the lazily loaded profiles.
+                _lazyPortableProfiles = new Lazy<CompiledNetPortableProfileCollection>(() => new CompiledNetPortableProfileCollection(BuildPortableProfileCollection()));
             }
         }
 
-        private static void CreateOptionalFrameworksDictionary()
-        {            
-            if (_portableProfiles != null)
+        private static IDictionary<string, NetPortableProfile> CreatePortableProfilesByCustomProfileString(NetPortableProfileCollection profileCollection)
```
Comment (alpaix) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: The method solely is consumed by `CompiledNetPortableProfileCollection` constructor. Does it make sense to move it there then?

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 2]
```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
...
+        {
+            return profileCollection.ToDictionary(x => x.CustomProfileString);
+        }
+
+        private static IDictionary<string, List<VersionStringISetTuple>> CreateOptionalFrameworksDictionary(NetPortableProfileCollection profileCollection)
```
Comment (alpaix) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: The same.

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 3]
```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
...
-            _portableProfilesByCustomProfileString.TryGetValue(profileName, out result);
+            Compiled.PortableProfilesByCustomProfileString.TryGetValue(profileName, out result);
 
             return result;
         }
 
         internal static NetPortableProfileCollection Profiles
         {
-            get
+            set
```
Comment (alpaix) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: set-only attribute looks weird. :confused: 
Replace with a method?

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 4]
```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
...
-            }
-            set
-            {
-                // This setter is only for Unit Tests.
-                _portableProfiles = value;
-                _portableProfilesByCustomProfileString = _portableProfiles.ToDictionary(x => x.CustomProfileString);
-                CreateOptionalFrameworksDictionary();
+                // This setter is only for unit tests and is NOT thread-safe.
+                // Reset the lazily loaded profiles.
+                _lazyPortableProfiles = new Lazy<CompiledNetPortableProfileCollection>(() => new CompiledNetPortableProfileCollection(BuildPortableProfileCollection()));
```
Comment (alpaix) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: why is this needed? Doesn't member initializer do the same thing?

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 5]
```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
...
+                    _portableProfiles = null;
+                }
+                else
                 {
-                    // We use the setter so that we can consistently set both the 
-                    // existing collection as well as the CustomProfileString-indexed one.
-                    // This keeps both in sync.
-                    Profiles = BuildPortableProfileCollection();
-                    CreateOptionalFrameworksDictionary();
+                    _portableProfiles = new CompiledNetPortableProfileCollection(value);
```
Comment (alpaix) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: you can change it to 

``` c#
_lazyPortableProfiles = new Lazy<CompiledNetPortableProfileCollection>(() => new CompiledNetPortableProfileCollection(value));
```

Then you can get rid of this `_portableProfiles` duality nonsense.

File: src/Core/NETPortable/NetPortableProfileTable.cs [hunk 6]
```diff
@@ -14,95 +13,95 @@ public static class NetPortableProfileTable
...
-            if (String.IsNullOrEmpty(profileName))
+            if (string.IsNullOrEmpty(profileName))
             {
                 throw new ArgumentException(CommonResources.Argument_Cannot_Be_Null_Or_Empty, "profileName");
             }
 
             // Original behavior fully preserved, as we first try the original behavior.
             // NOTE: this could be a single TryGetValue if this collection was kept as a dictionary...
-            if (Profiles.Contains(profileName))
+            if (Compiled.Profiles.Contains(profileName))
```
Comment (alpaix) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: That's not thread-safe at all (inconsistent). Each time in 3 member access occurrences here you may get different instances. Consider moving the method to the private class or cache the instance at the beginning to a local variable.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 3 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Sure

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 6 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Fixed.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 5 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Fixed.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 4 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Fixed.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 3 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Fixed.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 2 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Fixed.

File: src/Core/NETPortable/NetPortableProfileTable.cs (hunk 1 above)
Comment (joelverhagen) [inline src/Core/NETPortable/NetPortableProfileTable.cs]: Fixed.

**Comment by alpaix** [comment]
:cool: :shipit:
